jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -0,0 +1,343 @@
+     */
+    public static void copy(Content.Source source, Content.Sink sink, Predicate<Throwable> onTransientError, Callback callback)
+    {
+        new ContentCopierIteratingCallback(source, sink, onTransientError, callback).iterate();
+    }
+
+    /**
+     * {@link IteratingCallback} implementation that performs a copy from a {@link Content.Source} to a {@link Content.Sink}.
+     */
+    private static class ContentCopierIteratingCallback extends IteratingCallback
+    {
+        private final Content.Source source;
+        private final Content.Sink sink;
+        private final Predicate<Throwable> onTransientError;
+        private final Callback callback;
+        private boolean lastWritten;
+
+        public ContentCopierIteratingCallback(Content.Source source, Content.Sink target, Predicate<Throwable> onTransientError, Callback callback)
+        {
+            this.source = source;
+            this.sink = target;
+            this.onTransientError = onTransientError;
+            this.callback = callback;
+        }
+
+        @Override
+        protected Action process() throws Throwable
+        {
+            if (lastWritten)
+                return Action.SUCCEEDED;
+
+            Content.Chunk chunk = source.read();
+            if (chunk == null)
+            {
+                source.demand(this::succeeded);
+                return Action.SCHEDULED;
+            }
+            if (Content.Chunk.isFailure(chunk, false))
+            {
+                Throwable failure = chunk.getFailure();
+                if (onTransientError.test(failure))
+                    throw new IOException(failure);
+            }
+            if (Content.Chunk.isFailure(chunk, true))
+                throw new IOException(chunk.getFailure());
+
+            if (chunk.hasRemaining())
+            {
+                ByteBuffer byteBuffer = chunk.getByteBuffer();
+                sink.write(chunk.isLast(), byteBuffer, Callback.from(chunk::release, this));
+                lastWritten = chunk.isLast();
+                return Action.SCHEDULED;
+            }
+            else if (chunk.isLast())
+            {
+                sink.write(true, BufferUtil.EMPTY_BUFFER, Callback.from(chunk::release, this));
+                lastWritten = true;
+                return Action.SCHEDULED;
+            }
+            else
+            {
+                chunk.release();
+                return Action.SCHEDULED;
+            }
+        }
+
+        @Override
+        protected void onCompleteSuccess()
+        {
+            callback.succeeded();
+        }
+
+        @Override
+        protected void onCompleteFailure(Throwable x)
+        {
+            source.fail(x);
+            callback.failed(x);
+        }
+    }
+
+    /**
+     * <p>A specialized {@link PathContentSource}
+     * whose content is sliced by a byte range.</p>
+     */
+    private static class RangedPathContentSource extends PathContentSource
```

Comment:
There was identical functionality in `MultiPartByteRanges`, but with bits left out such as `ByteRange`.
Why are we moving this to `oej.io`, as byte ranges seems to be just an HTTP concept?

Reply:
True, `ByteRange` is an HTTP concept but it has to be implemented at the IO layer, and this PR is about centralizing IO concepts related to Resource so I think it's fair that the implementation moved to `IOResources`.